types/quick-store/index.d.ts

```diff
@@ -0,0 +1,25 @@
+// Type definitions for quick-store 0.1
+// Project: https://www.npmjs.com/package/quick-store (repository is deleted)
+// Definitions by: Glenn <https://github.com/promise>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export = QuickStore;
+
+interface dbContents { [key: string]: value; }
+type value = string | number | dbContents | value[] | boolean | null;
+type callback<value> = (data: value) => any;
+type callbackExists<optionalCallback, returnIfExists, returnIfNoExists> = optionalCallback extends callback<any> ? returnIfExists : returnIfNoExists;
+
+declare function QuickStore(filename?: string, data?: dbContents): QuickStore.Database;
+
+declare namespace QuickStore {
+    interface Database {
+        put(data: dbContents, callback?: callback<dbContents>): void;
+        setItem(key: string, value: value, callback?: callback<dbContents>): void;
+        getItem(key: string, callback: callback<value>): void;
+        removeItem(key: string, callback?: callback<dbContents>): void;
+        clear(callback?: callback<dbContents>): void;
+        get<optionalCallback extends callback<dbContents> | undefined>(callback?: optionalCallback): callbackExists<optionalCallback, void, dbContents>;
```

Comment:
you can use methods overloading  here I think, to denote different outcome. The rule is more specific on the top:
```ts
get(callback: CallbackType...): void; 
get(): Value
```
it's much cleaner to read and grasp that method actually has two different paths.
I can see usage of using TS features to actually create outcome type for return, but that usually make more sense in much more complicated declaration

Reply:
I tried doing this at first but got errors. Seems like it's working now, so probably something I did wrong yesterday. Thank you 👍